protocol/linkpreview/linkpreview.go

```diff
@@ -0,0 +1,321 @@
+			url: url.String(),
+			err: errors.New(""),
+		}
+	}
+
+	if ogMetadata.ThumbnailURL != "" {
+		t, err := fetchThumbnail(u.logger, ogMetadata.ThumbnailURL)
+		if err != nil {
+			// Given we want to fetch thumbnails on a best-effort basis, if an error
+			// happens we simply log it.
+			u.logger.Info("failed to fetch thumbnail", zap.String("url", url.String()), zap.Error(err))
+		} else {
+			preview.Thumbnail = t
+		}
+	}
+
+	preview.Title = ogMetadata.Title
+	preview.Description = ogMetadata.Description
+	return preview, nil
+}
+
+func newUnfurler(logger *zap.Logger, url *neturl.URL) Unfurler {
+	u := new(OpenGraphUnfurler)
+	u.logger = logger
+	return u
+}
+
+func unfurl(logger *zap.Logger, url string) (common.LinkPreview, error) {
+	var preview common.LinkPreview
+
+	parsedURL, err := neturl.Parse(url)
+	if err != nil {
+		return preview, err
+	}
+
+	unfurler := newUnfurler(logger, parsedURL)
+	preview, err = unfurler.unfurl(parsedURL)
+	if err != nil {
+		return preview, err
+	}
+	preview.Hostname = strings.ToLower(parsedURL.Hostname())
+
+	return preview, nil
+}
+
+// parseValidURL is a stricter version of url.Parse that performs additional
+// checks to ensure the URL is valid for clients to request a link preview.
+func parseValidURL(rawURL string) (*neturl.URL, error) {
+	u, err := neturl.Parse(rawURL)
+	if err != nil {
+		return nil, fmt.Errorf("parsing URL failed: %w", err)
+	}
+
+	if u.Scheme == "" {
+		return nil, errors.New("missing URL scheme")
+	}
+
+	_, err = publicsuffix.EffectiveTLDPlusOne(u.Hostname())
+	if err != nil {
+		return nil, fmt.Errorf("missing known URL domain: %w", err)
+	}
+
+	return u, nil
+}
+
+// GetURLs returns only what we consider unfurleable URLs.
+func GetURLs(text string) []string {
+	parsedText := markdown.Parse([]byte(text), nil)
+	visitor := common.RunLinksVisitor(parsedText)
+
+	urls := make([]string, 0, len(visitor.Links))
+	indexed := make(map[string]any, len(visitor.Links))
+
+	for _, rawURL := range visitor.Links {
+		parsedURL, err := parseValidURL(rawURL)
+		if err != nil {
+			continue
+		}
+		// Lowercase the host so the URL can be used as a cache key. Particularly on
+		// mobile clients it is common that the first character in a text input is
+		// automatically uppercased. In WhatsApp they incorrectly lowercase the
+		// URL's path, but this is incorrect. For instance, some URL shorteners are
+		// case-sensitive, some websites encode base64 in the path, etc.
+		parsedURL.Host = strings.ToLower(parsedURL.Host)
+
+		idx := parsedURL.String()
+		// Removes the spurious trailing forward slash.
+		idx = strings.TrimRight(idx, "/")
+		if _, exists := indexed[idx]; exists {
+			continue
+		} else {
+			indexed[idx] = nil
+			urls = append(urls, idx)
+		}
+	}
+
+	return urls
+}
+
+// UnfurlURLs assumes clients pass URLs verbatim that were validated and
+// processed by GetURLs.
+func UnfurlURLs(urls []string) ([]common.LinkPreview, error) {
+	logger, err := zap.NewDevelopment()
```

Comment:
I didn't quite understand @cammellos. In this part I had to instantiate a zap logger and I passed it around. Should I do it differently?

Reply:
Leaving the comment for reference, as @cammellos explained to me. It is important to get the logger instance from `api.service.messenger`, that way, logs are written to `geth.log`, otherwise they go to stdout only.

Done https://github.com/status-im/status-go/pull/3471/commits/a4039c6c0a1467a4425f419cd4321c358b9ef99c